### ISSUE TITLE
cli: applying migrations in a different "environment" after config v3 update fails

### ISSUE THREAD

**Issue by scriptonist** [hasura/graphql-engine#6861]
Scenario:

I've hasura in 2 environment Dev & Staging

My CLI project is in config v2 and I decided to update to config V3. Since I was in development. I ran `scripts update-project-v3` command on my dev hasura instance. My CLI project was updated successfully and migrations state was updated in my dev hasura instance.

Now, I did my dev and I want to promote it to staging.

I run command `hasura migrate apply --database-name default --endpoint <staging-endpoint>`. But this results in an error, because the state store for migrations are different in config v2 and config v3. 

Since the project is in config v3 it uses the new statestore, but this is empty in staging (state will get copied over only during `update-project-v3` command). So it tries to apply all migrations present. Which will result in applying migrations which might be already applied -- resulting in an error.

Possible solution: introduce a new flag which copies state from prev state store to new state store. `hasura scripts update-project-v3 --move-state-only`. This command can be run on an updated config v3 project to just copy over state.

**Comment by scriptonist** [comment]
fixed in https://github.com/hasura/graphql-engine/commit/9dfed5da5773a530c2ed2d9936fdf236cb456900

**Comment by mikejcooper** [comment]
@scriptonist Facing a somewhat similar issue. Working on v3 config, getting an 'already exists error' when trying to apply migrations to a copied db and hasura instance.  

The workflow is the following, on a PR: 

1. Beta database (containing public, hdb_catalog schema) is copied. 
2. New Hasura instance is created and copied db attached.
3. Run` hasura migrate apply...` on new hasura instance 

This results in the 'table already exist errors' etc. Implying the previous migration, applied to the beta db, has not been recognised.

I'm a bit confused about how Hasura is recording migration state in v3? I believe that in v2 it was recorded in the hdb_catalog schema? Could you shed some light on how this works?

**Comment by scriptonist** [comment]
@mikejcooper v2 CLI uses a server API to store the state (pre v2.0 this used to be `hdb_catalog.schema_migrations`). So if you were upgrading your project from config V2 -> config v3 using `hasura scripts update-project-v3` command, this migrations "state" transfer should have happened. 

The following is the API request CLI makes use of to fetch migrations state in the new paradigm.

```
curl --location --request POST 'http://localhost:8080/v1/metadata' \
--header 'x-hasura-admin-secret: XXXXX' \
--header 'Content-Type: application/json' \
--data-raw '{
    "type": "get_catalog_state",
    "args": {}
}'
```
Can you paste in the output of
```
hasura migrate status
```
and also the result of running the following SQL 

```
select * from hdb_catalog.schema_migrations;
```

**Comment by frandiox** [comment]
@scriptonist So I have a similar issue here. I've updated my local dev environment successfully. Now I'm moving my staging environment to Hasura Cloud. DB and all of that is connected. The only problem is when running `hasura migrate` because it tries to apply all the migrations that are already in place.
I've tried with `--move-state-only` and it succeeds. However, `hasura migrate` keeps failing after that.

The result from the `curl` command you wrote:

```
{"id":"01f6115b-facc-4125-868e-b284f5418abf","cli_state":{"settings":{"migration_mode":"true"},"isStateCopyCompleted":true},"console_state":{"console_notifications":{"admin":{"read":"default","showBadge":false,"date":"2021-07-08T14:18:40.926Z"}}}}% 
```

The status using the CLI is `Source Status - Present`, `Database Status - Not Present` for every migration.

The table `hdb_catalog.schema_migrations` is empty (exists, but empty).

How should I inform Hasura Cloud about the existing migrations in DB?
